R/get_analysis_date.R

```diff
@@ -0,0 +1,270 @@
+#'
+#' # simulate data
+#' temp <- simfix2simpwsurv(fail_rate)        # transfer the failure rate
+#' set.seed(2023)
+#' simu_data <- sim_pw_surv(
+#'   n = n,                                                            # sample size
+#'   stratum = tibble(stratum = stratum, p = prevelance_ratio),        # stratified design
+#'                                                                     # with prevalence ratio of 6:4
+#'   block =  c("control", "control", "experimental", "experimental"), # randomization ratio
+#'   enroll_rate = enroll_rate,                                        # enrollment rate
+#'   fail_rate = temp$fail_rate,                                       # failure rate
+#'   dropout_rate = temp$dropout_rate)                                 # dropout rate
+#'
+#' # example 1: cut for analysis at the 24-th month.
+#' get_analysis_date(
+#'   simu_data,
+#'   planned_calendar_time = 24)
+#'
+#' # example 2: cut for analysis when there are 300 events
+#' # in the overall population.
+#' get_analysis_date(
+#'   simu_data,
+#'   target_event_overall = 300)
+#'
+#' # example 3: cut for analysis at the 24-th month,
+#' # and there are 300 events in the overall population,
+#' # which arrives later.
+#' get_analysis_date(
+#'   simu_data,
+#'   planned_calendar_time = 24,
+#'   target_event_overall = 300)
+#'
+#' # example 4: cut for analysis when there are at least 100
+#' # events in the biomarker-positive population,
+#' # and at least 200 events in the biomarker-negative population,
+#' # which arrives later.
+#' get_analysis_date(
+#'   simu_data,
+#'   target_event_per_stratum = c(100, 200))
+#' get_analysis_date(
+#'   simu_data,
+#'   target_event_overall = 150,
+#'   target_event_per_stratum = c(100, NA))
+#'
+#' # example 5: cut for analysis when there are at least 100 events
+#' # in the biomarker positive population,
+#' # and at least 200 events in the biomarker negative population, which arrives later.
+#' # But will stop at the 30-th months if events less than 100/200.
+#' get_analysis_date(
+#'   simu_data,
+#'   target_event_per_stratum = c(100, 200),
+#'   max_extension_for_target_event = 30)
+#'
+#' # example 6: cut for analysis after 12 months followup when 80%
+#' # of the patients are enrolled in the overall population.
+#' get_analysis_date(
+#'   simu_data,
+#'   enroll_rate = enroll_rate,
+#'   min_n_overall = n * 0.8,
+#'   min_followup = 12)
+#'
+#' # example 7: cut for analysis when 12 months after at least
+#' # 200/160 patients are enrolled at the
+#' # biomarker positive/negative population.
+#' get_analysis_date(
+#'   simu_data,
+#'   enroll_rate = enroll_rate,
+#'   min_n_per_stratum = c(200, 160),
+#'   min_followup = 12)
+#' get_analysis_date(
+#'   simu_data,
+#'   enroll_rate = enroll_rate,
+#'   min_n_per_stratum = c(200, NA),
+#'   min_followup = 12)
+get_analysis_date <- function(
+    simu_data,
+    # Option 1: planned calendar time for the analysis
+    planned_calendar_time = NA,
```

Comment:
It is traditional to specify optional arguments as `NULL`

Reply:
Hello John, I appreciate your 2 valuable suggestions. I chose to use `NA` instead of `NULL` in this scenario to facilitate the implementation of `na.rm = TRUE` while using `pmin()`/`pmax()`. To expedite the development process, I temporarily used `NA`, and once its functionality is approved by Keaven/Yilong, I will hand over the responsibility of refining the code (like `NA` vs. `NULL`, and nested help functions) to you. Thanks!